src/router_core/transfer.c

```diff
@@ -413,6 +413,43 @@ static void qdr_delete_delivery_internal_CT(qdr_core_t *core, qdr_delivery_t *de
 }
 
 
+void qdr_delivery_link_peers_CT(qdr_delivery_t *in_dlv, qdr_delivery_t *out_dlv)
+{
+    if (in_dlv->peer || out_dlv->peer)
+        // One of the deliveries already has a peer. Don't proceed.
+        return;
+
+    out_dlv->peer = in_dlv;
+    in_dlv->peer = out_dlv;
+
+    qdr_delivery_incref(out_dlv);
+    qdr_delivery_incref(in_dlv);
+}
+
+
+void qdr_delivery_unlink_peers_CT(qdr_core_t *core, qdr_delivery_t *dlv, qdr_delivery_t *peer)
+{
+    dlv->peer  = 0;
+    peer->peer = 0;
+
+    qdr_delivery_decref_CT(core, dlv);
```

Comment:
What if these two deliveries weren't linked to one another?  You will cause a data corruption with no indication.  You should assert that the proper mutual linkage is in place before removing it and decrementing reference counts.